keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -613,7 +621,178 @@ sequence that it will not successfully decode.
+| 6  | dynamic | static  | request   | cbor     |
+| 7  | dynamic | static  | response  | json     |
+| 8  | dynamic | static  | response  | cbor     |
+| 9  | static  | dynamic | request   | json     |
+| 10 | static  | dynamic | request   | cbor     |
+| 11 | static  | dynamic | response  | json     |
+| 12 | static  | dynamic | response  | cbor     |
+| 13 | static  | static  | request   | protobuf |
+| 14 | static  | static  | response  | protobuf |
+
+In cases 5-6 and 11-12, the marshalling side acts on an Unstructured object and is not aware that
+the unmarshalling side may decode some of the payload into a `RawExtension`. The bytes stored in the
+`RawExtension` by unmarshalling ultimately depend on the negotiated content type, which can vary
+with the enablement of the CBOR serializer. Existing programs have so far been able to assume that
+unmarshalled RawExtensions always have either nil or a valid JSON text in their Raw field.
+
+In cases 1-4, 7-8, and 9-10, if the marshalling side populates Raw with a non-nil slice, it is
+responsible for ensuring that that encoding of the slice contents matches the encoding that will be
+used to serialize the object containing the `RawExtension`. This is trivially ensured in cases 13
+and 14 because Protobuf is capable of representing `RawExtension` values containing arbitrary
+bytes. Protobuf is not a supported encoding for Unstructured objects. Existing programs have in
+practice stored JSON in the Raw field of `RawExtension`.
+
+#### Compatibility
+
+If the `RawExtension` marshalling and unmarshalling behavior for CBOR were to be implemented in
+exactly the same way as the existing JSON behaviors, the assumptions in many existing programs that
+the Raw field can be assigned to a slice of JSON bytes, or that the Raw bytes of an unmarshalled
+`RawExtension` are valid JSON, would be broken.
+
+The simple approach of automatically transcoding JSON to CBOR during CBOR marshalling, and
+transcoding CBOR to JSON during CBOR unmarshalling, would avoid breaking existing programs. However,
+the expense of transcoding to or from JSON would negate any performance advantage of a binary
+encoding. This expense would not be limited to a few API types: significant examples include the use
+of a `RawExtension` field in `metav1.WatchEvent` to represent each watch event's object state, or
+the arbitrary objects embedded in `admissionv1.AdmissionRequest`.
+
+Instead, automatic transcoding to and from JSON will become optional and initially enabled by
+default.
+
+For marshalling:
+
+1. If both Raw and Object are `nil`, `null` is returned.
+1. If Object is not `nil`, return the result of marshalling Object.
+1. If Raw is prefixed by the CBOR encoding of the head of the "self-described CBOR" tag, then return
+   the Raw bytes verbatim. This is the same prefix used by the CBOR `recognizer.RecognizingDecoder`
+   implementation, and all objects encoded by the CBOR `runtime.Encoder` implementation have this
+   prefix.
```

Comment:
Is this saying that CBOR data is always possible to identify by just checking the first few bytes?  This is great.  Are we saying here that we'll always be able to marhal RawExtension containing either json or cbor because we can tell the difference by just looking at the data?